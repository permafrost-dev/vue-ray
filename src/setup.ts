import { RayMixin } from '@/RayMixin';
import type { ray } from '@/VueRay';
import { rayWrapped } from '@/VueRay';
import type { Ref } from 'vue';
import { getCurrentInstance, onBeforeMount, onBeforeUnmount, onMounted, onUnmounted, onUpdated, ref } from 'vue';

export type RaySetupLifecycleEventOptions = {
    beforeMount?: boolean;
    beforeUnmount?: boolean;
    updated?: boolean;
    mounted?: boolean;
    unmounted?: boolean;
    all?: boolean;
};

export type RaySetupOptions = {
    lifecycleEvents?: RaySetupLifecycleEventOptions;
    lifecycleMethods?: RaySetupOptionsVueLifecycleMethods;
};

export type RaySetupOptionsVueLifecycleMethods = {
    beforeMount?: typeof onBeforeMount;
    beforeUnmount?: typeof onBeforeUnmount;
    updated?: typeof onUpdated;
    mounted?: typeof onMounted;
    unmounted?: typeof onUnmounted;
};

export function raySetup(component = null, options: RaySetupOptions = {}): Ref<typeof ray> {
    component = component ?? getCurrentInstance();

    const $ray = rayWrapped(component);

    component.$ray = $ray;

    options.lifecycleEvents = options.lifecycleEvents ?? {};
    options.lifecycleEvents.all = options.lifecycleEvents?.all ?? false;

    if (options.lifecycleEvents.all === true) {
        options.lifecycleEvents.beforeMount = true;
        options.lifecycleEvents.beforeUnmount = true;
        options.lifecycleEvents.updated = true;
        options.lifecycleEvents.mounted = true;
        options.lifecycleEvents.unmounted = true;
    }

    if (options.lifecycleEvents.beforeMount) {
        const callback = options.lifecycleMethods?.beforeMount ?? onBeforeMount;
        callback(() => RayMixin.beforeMount(component, $ray()));
    }

    if (options.lifecycleEvents.beforeUnmount) {
        const callback = options.lifecycleMethods?.beforeUnmount ?? onBeforeUnmount;
        callback(() => RayMixin.beforeUnmount(component, $ray()));
    }

    if (options.lifecycleEvents.updated) {
        const callback = options.lifecycleMethods?.updated ?? onUpdated;
        callback(() => RayMixin.updated(component, $ray()));
    }

    if (options.lifecycleEvents.mounted) {
        const callback = options.lifecycleMethods?.mounted ?? onMounted;
        callback(() => RayMixin.mounted(component, $ray()));
    }

    if (options.lifecycleEvents.unmounted) {
        const callback = options.lifecycleMethods?.unmounted ?? onUnmounted;
        callback(() => RayMixin.unmounted(component, $ray()));
    }

    return ref(component.$ray);
}
