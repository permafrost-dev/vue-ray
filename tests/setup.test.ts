import { raySetup } from '@/setup';
import { expect, it } from 'vitest';
import { isRef } from 'vue';

function setupCallCounters() {
    const component = { data: {} } as any;
    const callCounters = {
        beforeMount: 0,
        beforeUnmount: 0,
        updated: 0,
        mounted: 0,
        unmounted: 0,
    };

    const lifecycleMethods = {
        beforeMount: (hook): false => {
            callCounters.beforeMount++;
            hook();
            return false;
        },
        beforeUnmount: (hook): false => {
            callCounters.beforeUnmount++;
            hook();
            return false;
        },
        mounted: (hook): false => {
            callCounters.mounted++;
            hook();
            return false;
        },
        unmounted: (hook): false => {
            callCounters.unmounted++;
            hook();
            return false;
        },
        updated: (hook): false => {
            callCounters.updated++;
            hook();
            return false;
        },
    };

    return {
        component,
        callCounters,
        lifecycleMethods,
    };
}

it('performs setup for ray and returns a ref', () => {
    const component = { data: {} } as any;
    const result = raySetup(component);

    expect(result.value).not.toBeNull();
    expect(isRef(result)).toBe(true);
});

it('adds the ray methods to the component', () => {
    const component = { data: {} } as any;
    raySetup(component);

    expect(component.$ray).toBeDefined();
});

it('adds all lifecycle events to the component', () => {
    const { component, callCounters, lifecycleMethods } = setupCallCounters();

    raySetup(component, {
        lifecycleEvents: {
            all: true,
        },
        lifecycleMethods,
    });

    expect(callCounters).toMatchSnapshot();
});

it('adds beforeMount lifecycle event to the component', () => {
    const { component, callCounters, lifecycleMethods } = setupCallCounters();

    raySetup(component, {
        lifecycleEvents: {
            beforeMount: true,
        },
        lifecycleMethods,
    });

    expect(callCounters).toMatchSnapshot();
});

it('adds beforeUnmount lifecycle event to the component', () => {
    const { component, callCounters, lifecycleMethods } = setupCallCounters();

    raySetup(component, {
        lifecycleEvents: {
            beforeUnmount: true,
        },
        lifecycleMethods,
    });

    expect(callCounters).toMatchSnapshot();
});

it('adds updated lifecycle event to the component', () => {
    const { component, callCounters, lifecycleMethods } = setupCallCounters();

    raySetup(component, {
        lifecycleEvents: {
            updated: true,
        },
        lifecycleMethods,
    });

    expect(callCounters).toMatchSnapshot();
});

it('adds mounted lifecycle event to the component', () => {
    const { component, callCounters, lifecycleMethods } = setupCallCounters();

    raySetup(component, {
        lifecycleEvents: {
            mounted: true,
        },
        lifecycleMethods,
    });

    expect(callCounters).toMatchSnapshot();
});

it('adds unmounted lifecycle event to the component', () => {
    const { component, callCounters, lifecycleMethods } = setupCallCounters();

    raySetup(component, {
        lifecycleEvents: {
            unmounted: true,
        },
        lifecycleMethods,
    });

    expect(callCounters).toMatchSnapshot();
});
