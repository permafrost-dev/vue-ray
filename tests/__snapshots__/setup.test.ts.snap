// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`adds all lifecycle events to the component 1`] = `
{
  "beforeMount": 1,
  "beforeUnmount": 1,
  "mounted": 1,
  "unmounted": 1,
  "updated": 1,
}
`;

exports[`adds beforeMount lifecycle event to the component 1`] = `
{
  "beforeMount": 1,
  "beforeUnmount": 0,
  "mounted": 0,
  "unmounted": 0,
  "updated": 0,
}
`;

exports[`adds beforeUnmount lifecycle event to the component 1`] = `
{
  "beforeMount": 0,
  "beforeUnmount": 1,
  "mounted": 0,
  "unmounted": 0,
  "updated": 0,
}
`;

exports[`adds mounted lifecycle event to the component 1`] = `
{
  "beforeMount": 0,
  "beforeUnmount": 0,
  "mounted": 1,
  "unmounted": 0,
  "updated": 0,
}
`;

exports[`adds unmounted lifecycle event to the component 1`] = `
{
  "beforeMount": 0,
  "beforeUnmount": 0,
  "mounted": 0,
  "unmounted": 1,
  "updated": 0,
}
`;

exports[`adds updated lifecycle event to the component 1`] = `
{
  "beforeMount": 0,
  "beforeUnmount": 0,
  "mounted": 0,
  "unmounted": 0,
  "updated": 1,
}
`;
